crates/core_simd/src/vector.rs

```diff
@@ -979,3 +1189,37 @@ where
 {
     type Mask = isize;
 }
+
+#[inline]
+fn lane_indices<const N: usize>() -> Simd<usize, N>
+where
+    LaneCount<N>: SupportedLaneCount,
+{
+    let mut index = [0; N];
+    for i in 0..N {
+        index[i] = i;
+    }
+    Simd::from_array(index)
+}
+
+#[inline]
+fn mask_up_to<M, const N: usize>(len: usize) -> Mask<M, N>
+where
+    LaneCount<N>: SupportedLaneCount,
+    M: MaskElement,
+{
+    let index = lane_indices::<N>();
+    let max_value: u64 = M::max_unsigned();
+    macro_rules! case {
+        ($ty:ty) => {
+            if N < <$ty>::MAX as usize && max_value as $ty as u64 == max_value {
+                return index.cast().simd_lt(Simd::splat(len.min(N) as $ty)).cast();
+            }
+        };
+    }
+    case!(u8);
+    case!(u16);
+    case!(u32);
+    case!(u64);
+    index.simd_lt(Simd::splat(len)).cast()
+}
```

Comment:
How about adding a `from_usize` and doing something like this to avoid casts?
```suggestion
#[inline]
fn lane_indices<M, const N: usize>() -> Simd<M, N>
where
    M: MaskElement,
    LaneCount<N>: SupportedLaneCount,
{
    Simd::from_array(core::array::from_fn(M::from_usize))
}

#[inline]
fn mask_up_to<M, const N: usize>(len: usize) -> Mask<M, N>
where
    LaneCount<N>: SupportedLaneCount,
    M: MaskElement,
{
    if N as u64 < M::max_unsigned() {
        lane_indices::<M, N>().simd_lt(Simd::splat(M::from_usize(len)))
    } else {
        lane_indices::<usize, N>().simd_lt(Simd::splat(len)).cast()
    }
}
```

Reply:
part of the `case!` logic is picking the minimum viable element size, since that's usually more efficient, so it would pick 16-bit elements for `Simd<u8, 256>` and 32-bit elements for `Simd<u8, 65536>` (yes, afaict that is theoretically possible in a single vector on RISC-V V).

your suggestion would just pick the mask's element size and if that isn't big enough, give up and use full `usize` elements even if 16-bit elements would suffice for all realistic vector types. Unfortunately LLVM doesn't seem to be able to optimize vectors to use a smaller element size unless you explicitly cast to that size before using `simd_lt`

Reply:
also `lane_indices::<M, N>().simd_lt(Simd::splat(M::from_usize(len)))` is just plain incorrect, if `M` is `i8` and `len` is `256` then it will wrap around (or panic if `from_usize` checks) and act as if `len == 0`

Reply:
I think for the usual case (vectors with well under 256 elements) codegen is likely to be better without a cast?  How about keeping as is, but trying to use M first if it fits?

Reply:
maybe? the casts optimized away completely when I tried...

Reply:
Interesting, I've had bad luck with casts, but that's usually with more complex operations. If we think the casts won't matter it's fine with me

Reply:
> Would it be sufficient to consider only value of N when determine the element size? i.e. use 8-bit elements for `Simd<u64,16>`

no, because iirc LLVM doesn't optimize-out the element-size conversions then.